src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -142,6 +147,74 @@ public async Task StopContainersAsync(IExecutionContext executionContext, object
+            }
+            var accessToken = await credential.GetTokenAsync(new TokenRequestContext(new[] { "https://management.core.windows.net/" }), cancellationToken);
+            // To print the token, you can convert it to string 
+            return accessToken.Token.ToString();
+        }
+
+        private async Task<string> GetACRPasswordFromAADToken(string AADToken, string tenantId, string registryServer, int retryCount, int timeToWait, CancellationToken cancellationToken)
+        {
+            Uri url = new Uri(registryServer + "/oauth2/exchange");
+            const int retryLimit = 5;
+            using var httpClientHandler = HostContext.CreateHttpClientHandler();
+            using HttpClient httpClient = new HttpClient(httpClientHandler);
+            httpClient.DefaultRequestHeaders.Add("Content-Type", "application/x-www-form-urlencoded");
+            using StringContent stringContent = new StringContent("grant_type=access_token&service=" + registryServer[8..] + "&tenant=" + tenantId + "&access_token=" + AADToken);
+            string ACRpassword = string.Empty;
+            int waitedTime = 0;
+            do
+            {
+                try
+                {
+                    var response = await httpClient.PostAsync(url, stringContent, cancellationToken).ConfigureAwait(false);
+                    if (response.StatusCode == HttpStatusCode.OK)
+                    {
+                        Dictionary<string, string> list = (Dictionary<string, string>)JsonConvert.DeserializeObject(await response.Content.ReadAsStringAsync(), typeof(Dictionary<string, string>));
+                        ACRpassword = list["refresh_token"];
+                    }
+                    else if (response.StatusCode == HttpStatusCode.TooManyRequests || response.StatusCode == HttpStatusCode.InternalServerError)
+                    {
+                        if (retryCount < retryLimit)
+                        {
+                            waitedTime = 2000 + timeToWait * 2;
+                            retryCount++;
+                            await Task.Delay(timeToWait);
+                        }
+                        else
+                        {
+                            throw new NotSupportedException("Could not fetch access token for ACR at the moment. Please try again in a few minutes.");
+                        }
+                    }
+                    else
+                    {
+                        throw new NotSupportedException("Could not fetch access token for ACR. Please configure Managed Service Identity (MSI) for Azure Container Registry with the appropriate permissions - https://stackoverflow.com/questions/71711086/pull-azure-container-registry-image-from-app-service-using-managed-identity.");
+                    }
+                }
+                catch
+                {
+                    waitedTime = 2000 + timeToWait * 2;
```

Comment:
Maybe move `waitedTime` calculation outside of the loop since `timeToWait` is passed in and shouldn't change